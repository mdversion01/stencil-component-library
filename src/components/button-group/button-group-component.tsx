// src/components/button-group/button-group.tsx
import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'button-group',
  styleUrl: 'button-group.scss',
  // shadow: false,
})
export class ButtonGroup {
  @Prop() ariaLabel: string = '';
  @Prop() classNames: string = '';
  @Prop() vertical: boolean = false;

  private getClassAttribute(): string {
    const baseClass = this.vertical ? 'btn-group-vertical' : 'btn-group';
    return [baseClass, this.classNames].filter(Boolean).join(' ');
  }

  render() {
    const classAttribute = this.getClassAttribute();

    return (
      <div
        role="group"
        aria-label={this.ariaLabel || 'Button Group'}
        class={classAttribute}
      >
        <slot />
      </div>
    );
  }
}
