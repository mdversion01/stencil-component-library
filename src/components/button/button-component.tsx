import { Component, Prop, Event, EventEmitter, h } from '@stencil/core';
import { logInfo } from '../../utils/log-debug';

@Component({
  tag: 'button-component',
  styleUrl: 'button.scss',
  // styleUrls: ['../../global/global.scss',],
})
export class Button {
  @Prop() absolute: boolean = false;
  @Prop() active: boolean = false;
  @Prop() ariaLabel: string = '';
  @Prop() block: boolean = false;
  @Prop() bottom: string = '';
  @Prop() btnIcon: boolean = false;
  @Prop() btnText: string = '';
  @Prop() classNames: string = '';
  @Prop() disabled: boolean = false;
  @Prop() end: boolean = false;
  @Prop() elevation: string = '';
  @Prop() fixed: boolean = false;
  @Prop() groupBtn: boolean = false;
  @Prop() iconBtn: boolean = false;
  @Prop({ reflect: true }) slotSide: 'left' | 'right' | 'none' = 'none';
  @Prop() styles: string = '';
  @Prop() left: string = '';
  @Prop() link: boolean = false;
  @Prop() outlined: boolean = false;
  @Prop() pressed: boolean | string = false;
  @Prop() right: string = '';
  @Prop() ripple: boolean = false;
  @Prop() shape: string = '';
  @Prop() size: string = '';
  @Prop() start: boolean = false;
  @Prop() stripped: boolean = false;
  @Prop() text: boolean = false;
  @Prop() textBtn: boolean = false;
  @Prop() titleAttr: string = '';
  @Prop() top: string = '';
  @Prop() url: string = '';
  @Prop() variant: string = '';
  @Prop() vertical: boolean = false;
  @Prop() zIndex: string = '';
  @Prop({ mutable: true }) isOpen: boolean = false;
  @Prop() targetId: string = '';
  @Prop() accordion: boolean = false;
  @Prop() devMode: boolean = false;

  @Event() customClick: EventEmitter;

  private handleClick() {
    this.customClick.emit();
    logInfo(this.devMode, 'Button', 'Clicked');
  }

  private handleKeyDown = (event: KeyboardEvent) => {
    if ((event.key === 'Enter' || event.key === ' ') && !this.disabled) {
      event.preventDefault();
      this.handleClick();
    }
  };

  private getDynamicStyles(): { [key: string]: string } {
    const styles: { [key: string]: string } = {};
    if (this.absolute) styles.position = 'absolute';
    if (this.fixed) styles.position = 'fixed';
    if (this.left) styles.left = `${this.left}px`;
    if (this.right) styles.right = `${this.right}px`;
    if (this.top) styles.top = `${this.top}px`;
    if (this.bottom) styles.bottom = `${this.bottom}px`;
    if (this.zIndex) styles['z-index'] = this.zIndex;

    logInfo(this.devMode, 'Button', 'getDynamicStyles()', styles);
    return styles;
  }

  private renderButtonContent() {
    const hasLeft = this.slotSide === 'left';
    const hasRight = this.slotSide === 'right';

    const content = this.btnIcon ? (
      <slot />
    ) : this.iconBtn ? (
      <span class="btn__content">
        <slot />
      </span>
    ) : (
      [
        hasLeft && <slot />,
        <span class={`btn__content ${hasLeft ? 'ms-1' : ''} ${hasRight ? 'me-1' : ''}`}>
          {this.btnText}
          {!hasLeft && !hasRight && <slot />}
        </span>,
        hasRight && <slot />,
      ]
    );

    logInfo(this.devMode, 'Button', 'renderButtonContent()', {
      btnText: this.btnText,
      slotSide: this.slotSide,
      btnIcon: this.btnIcon,
      iconBtn: this.iconBtn,
    });

    return content;
  }

  render() {
    const dynamicStyles = this.getDynamicStyles();
    const isGroup = !!this.groupBtn;
    const buttonGroup = this.vertical ? 'btn-group-vertical' : 'btn-group';
    const placement = this.start ? `${buttonGroup}-start` : this.end ? `${buttonGroup}-end` : 'btn-group__btn';

    const classList = [
      'btn',
      this.classNames,
      this.outlined ? `btn-outline-${this.variant}` : '',
      !this.outlined && this.variant ? `btn-${this.variant}` : '',
      this.block && 'btn--block',
      this.size && `btn-${this.size}`,
      this.shape,
      this.active && 'active',
      isGroup && buttonGroup,
      isGroup && placement,
      this.ripple && 'btn-ripple',
      this.stripped && 'stripped',
      this.iconBtn && 'icon-btn',
      this.link && 'link',
      this.elevation && `elevated-${this.elevation}`,
      this.textBtn && 'text-btn',
      this.text && 'text',
      this.btnIcon && 'btn-icon',
    ]
      .filter(Boolean)
      .join(' ');

    // Accessibility: ensure an aria-label when there is no visible text
    const hasVisibleText = !!(this.btnText && this.btnText.trim().length);
    const computedAriaLabel =
      (this.ariaLabel && this.ariaLabel.trim()) ||
      (hasVisibleText ? this.btnText.trim() : '') ||
      (this.titleAttr && this.titleAttr.trim()) ||
      'Button';

    if (!hasVisibleText && !this.ariaLabel && (this.btnIcon || this.iconBtn) && this.devMode) {
      console.warn(
        '[ButtonComponent] Icon-only button has no `ariaLabel`/text. Setting aria-label="%s".',
        computedAriaLabel
      );
    }

    logInfo(this.devMode, 'Button', 'render()', {
      classList,
      link: this.link,
      disabled: this.disabled,
      slotSide: this.slotSide,
      computedAriaLabel,
    });

    const ariaProps: Record<string, any> = {};
    if (this.accordion && this.targetId) {
      ariaProps['aria-expanded'] = String(this.isOpen);
      ariaProps['aria-controls'] = this.targetId;
    }

    const buttonProps = {
      ...ariaProps,
      'aria-pressed': this.pressed === 'true' || this.pressed === 'false' ? this.pressed : undefined,
      'aria-label': computedAriaLabel,
      'aria-disabled': this.link && this.disabled ? 'true' : undefined,
      style: dynamicStyles,
      class: classList,
      disabled: !this.link && this.disabled,
      onClick: (event: MouseEvent) => {
        if (this.disabled) {
          event.preventDefault();
          return;
        }
        const shouldPrevent = !this.url || this.url === '#' || this.url.trim() === '' || this.url.startsWith('#');
        if (this.link && shouldPrevent) event.preventDefault();
        this.handleClick();
      },
    };

    return this.link ? (
      <a
        {...(buttonProps as any)}
        href={this.url && this.url.trim() !== '' ? this.url : undefined}
        role="button"
        tabindex={this.disabled ? '-1' : '0'}
        onKeyDown={this.disabled ? undefined : this.handleKeyDown}
        title={this.titleAttr}
      >
        {this.renderButtonContent()}
      </a>
    ) : (
      <button {...(buttonProps as any)} type="button" title={this.titleAttr}>
        {this.renderButtonContent()}
      </button>
    );
  }
}
