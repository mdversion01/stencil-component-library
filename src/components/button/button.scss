@import '../badge/badge.scss';

// _btn-sizes.scss

$btn-sizes: (
  xs: (
    font: 0.625rem,
    height: 20px,
    min: 36px,
    padding: (
      0 8.8888888889px 2px,
    )
    // keep 3-value form,,,,,,,,
  ),
  sm: (
    font: 0.75rem,
    height: 28px,
    min: 32px,
    padding: (
      0 12.4444444444px,
    ),
  ),
  lg: (
    font: 1.2rem,
    height: 44px,
    min: 78px,
    padding: (
      0 19.5555555556px,
    ),
  ),
);

// why: centralize size tokens; easy to add/override
@mixin make-btn-size($name, $cfg, $prefix: '.btn-', $important: true) {
  #{$prefix}#{$name} {
    font-size: map-get($cfg, font) !important;
    height: map-get($cfg, height);
    min-width: map-get($cfg, min);
    $imp: if($important, ' !important', '');
    padding: #{map-get($cfg, padding)}#{$imp};
  }
}

// Generate .btn-xs / .btn-sm / .btn-lg
@each $name, $cfg in $btn-sizes {
  @include make-btn-size($name, $cfg);
}

// _btn-text-variants.scss

@mixin text-button-base($base-text: rgba(0, 0, 0, 0.87), $hover-bg: #e1e1e1) {
  .btn.text-btn,
  .btn.text,
  .btn.btn-icon {
    background-color: transparent;
    border: none;
    color: $base-text;
  }

  .btn.text-btn:hover,
  .btn.text:hover,
  .btn.btn-icon:hover {
    color: $base-text;
    background-color: $hover-bg;
  }
}

$btn-text-variants: (
  primary: (
    base: #2680eb,
    hover-bg: #146dd7,
    hover-border: #146dd7,
    hover-text: #f2f2f2,
    plain-hover: #146dd7,
  ),
  secondary: (
    base: #909090,
    hover-bg: #7d7d7d,
    hover-border: #7d7d7d,
    hover-text: #f2f2f2,
    plain-hover: #7d7d7d,
  ),
  tertiary: (
    base: #b9b9b9,
    hover-bg: #a6a6a6,
    hover-border: #a6a6a6,
    hover-text: #f2f2f2,
    plain-hover: #a6a6a6,
  ),
  success: (
    base: #2d9d78,
    hover-bg: #247f61,
    hover-border: #247f61,
    hover-text: #f2f2f2,
    plain-hover: #247f61,
  ),
  danger: (
    base: #e34850,
    hover-bg: #de2730,
    hover-border: #de2730,
    hover-text: #f2f2f2,
    plain-hover: #de2730,
  ),
  warning: (
    base: #e68619,
    hover-bg: #c47215,
    hover-border: #c47215,
    hover-text: #f2f2f2,
    plain-hover: #c47215,
  ),
  info: (
    base: #5aa9fa,
    hover-bg: #3596f9,
    hover-border: #3596f9,
    hover-text: #f2f2f2,
    plain-hover: #3596f9,
  ),
  light: (
    base: #f2f2f2,
    hover-bg: #dfdfdf,
    hover-border: #dfdfdf,
    hover-text: #1f1f1f,
    plain-hover: #dfdfdf,
  ),
  dark: (
    base: #323232,
    hover-bg: #505050,
    hover-border: #252525,
    hover-text: #f2f2f2,
    plain-hover: #505050,
  ),
  // NEW: neutral
  neutral:
    (
      base: #7a7a7a,
      hover-bg: #6a6a6a,
      hover-border: #6a6a6a,
      hover-text: #f2f2f2,
      plain-hover: #6a6a6a,
    ),
);

@mixin text-button-variant($name, $cfg) {
  $base: map-get($cfg, base);
  $hover-bg: map-get($cfg, hover-bg);
  $hover-border: map-get($cfg, hover-border);
  $hover-text: map-get($cfg, hover-text);
  $plain-hover: map-get($cfg, plain-hover);

  .btn.text-btn.btn-#{$name},
  .btn.text.btn-#{$name},
  .btn.btn-icon.btn-#{$name} {
    color: $base;
  }

  .btn.text-btn.btn-#{$name}:hover {
    color: $hover-text;
    background-color: $hover-bg;
    border-color: $hover-border;
  }

  .btn.text.btn-#{$name}:hover,
  .btn.btn-icon.btn-#{$name}:hover,
  .btn.text.btn-#{$name}:active,
  .btn.btn-icon.btn-#{$name}:active {
    color: $plain-hover;
    background: none;
    border: none;
  }
}

// Generate CSS
@include text-button-base();

@each $name, $cfg in $btn-text-variants {
  @include text-button-variant($name, $cfg);
}

.btn-flex {
  align-items: center;
  display: flex;
  justify-content: space-between;
}

.btn-ripple {
  display: inline-block;
  position: relative;
  overflow: hidden;
  transition: all ease-in-out 0.5s;

  &::after {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 25%;
    height: 100%;
    width: 50%;
    background-color: #000;
    border-radius: 30%;
    opacity: 0;
    pointer-events: none;
    transition: all ease-in-out 1s;
    transform: scale(5, 5);
  }

  &:active::after {
    padding: 0;
    margin: 0;
    opacity: 0.2;
    transition: 0s;
    transform: scale(0, 0);
  }
}

/* Button styles */
.btn {
  align-items: center;
  display: inline-flex !important;
  flex: 0 0 auto;
  letter-spacing: 0.0892857143em;
  justify-content: center;
  // outline: 0;
  position: relative;
}

// .btn:hover {
//   background-color: rgb(118, 118, 118);
//   color: #fff;
//   text-decoration: none;
// }

// .btn:before {
//   border-radius: inherit;
//   bottom: 0;
//   color: inherit;
//   content: '';
//   left: 0;
//   opacity: 0;
//   pointer-events: none;
//   position: absolute;
//   right: 0;
//   top: 0;
//   transition: opacity 0.2s cubic-bezier(0.4, 0, 0.6, 1);
// }

// .btn:hover:before {
//   opacity: 0.08;
// }

.btn-default {
  background-color: #e7e7e7 !important;
  border-color: #e7e7e7 !important;
  color: #212529 !important;

  &:hover {
    background-color: #767676 !important;
    border-color: #767676 !important;
    color: #fff !important;
  }

  &:focus {
    box-shadow: 0 0 0 0.2rem #767676;
  }

  &.active,
  &:not(:disabled):not(.disabled).active,
  &:not(:disabled):not(.disabled):active {
    background-color: #767676;
    border-color: #767676;
    color: #ffffff;
  }
}

.btn-outline-default {
  background-color: transparent !important;
  border-color: #e7e7e7 !important;
  color: #212529 !important;

  &:hover {
    background-color: #767676 !important;
    border-color: #767676 !important;
    color: #ffffff !important;
  }

  &:focus {
    box-shadow: 0 0 0 0.2rem #767676 !important;
  }

  &.active,
  &:not(:disabled):not(.disabled).active,
  &:not(:disabled):not(.disabled):active {
    border-color: #767676 !important;
    color: #fff !important;
  }
}

.btn.disabled,
.btn:disabled {
  opacity: 0.65;
}

.btn:focus {
  outline: 0;
  box-shadow: 0 0 0 0.2rem #8e8e8e;
}

.btn:focus-visible,
.btn-default:focus-visible {
  // color: var(--bs-btn-hover-color);
  // background-color: var(--bs-btn-hover-bg);
  // border-color: var(--bs-btn-hover-border-color);
  outline: 0 !important;
  box-shadow: 0 0 0 0.2rem #8e8e8e !important;
}

.btn.stripped {
  background-color: transparent !important;
  border-color: transparent !important;
  color: #333;
  height: auto !important;
  margin: 0 !important;
  padding: 0 !important;
  width: auto !important;

  &:hover {
    background-color: transparent !important;
    border-color: transparent !important;
    /* color: inherit !important; */
  }

  .btn--block {
    display: flex;
    flex: 1 0 auto;
    min-width: 100% !important;
    max-width: none;
  }
}

.btn__content {
  align-items: center;
  // color: inherit;
  display: flex;
  flex: 1 1 auto;
  justify-content: inherit;
  line-height: inherit;
  position: relative;
  transition: inherit;
  transition-property: opacity;
  width: 100%;
}

.btn.btn-active-blue {
  background-color: #909090 !important;
  border-color: #838383 !important;
  color: #252525;

  &.active,
  &:hover,
  &:active {
    background-color: #5aa9fa !important;
    color: #252525;
  }

  &.active {
    border-color: #419cf9;
  }
}

.btn.icon-btn.btn-active-blue.active {
  background-color: transparent;
  border-color: transparent;
  color: #419cf9;
}

.btn.size {
  align-items: center;
  border-radius: 0.25rem;
  display: inline-flex;
  font-size: 0.8333rem;
  font-weight: normal;
  height: 1.375rem;
  line-height: 1.2;
  margin: 0.125rem;
  padding: 0 0.5rem;
}

.btn.default,
.btn:not(.btn.circle).btn.default {
  height: 36px;
  min-width: 64px;
  padding: 0 16px;
}

.btn.btn-icon {
  height: auto !important;
  min-width: 10px !important;
  padding: 0 4px !important;
}

.btn:not(.btn--outlined).accent,
.btn:not(.btn-outlined).danger,
.btn:not(.btn-outlined).info,
.btn:not(.btn-outlined).primary,
.btn:not(.btn-outlined).secondary,
.btn:not(.btn-outlined).success,
.btn:not(.btn-outlined).warning,
.btn:not(.btn-outlined).dark {
  color: #fff;
}

.btn.pill {
  border-radius: 28px;
}

.btn.square {
  border-radius: 0;
}

.btn.circle {
  border-radius: 50%;
}

.btn {
  padding: 5px 10px !important;

  icon-component {
    .fa-ellipsis-vertical {
      font-size: 1.5rem;
    }
    .fa-gear {
      font-size: 1.25rem;
    }
  }
}

.btn.icon-btn {
  font-size: 1rem;
  height: 36px;
  width: 36px;
}

.btn.fab,
.btn.icon-btn {
  min-height: 0;
  min-width: 0;
  padding: 0;
}

.btn.icon-btn.regular-icon:hover {
  background-color: transparent;
  color: #4b9cf5;
}

.btn.icon-btn.icon-lg {
  font-size: 0.875rem !important;
  height: 44px !important;
  width: 44px !important;
}

.btn.icon-btn.large > i {
  font-size: 1.5rem !important;
}

.btn.icon-btn.icon-sm {
  font-size: 0.75rem !important;
  height: 28px !important;
  width: 28px !important;
}

.btn.icon-btn.small > i {
  font-size: 1rem !important;
}

.btn.icon-btn.icon-xs {
  font-size: 0.75rem !important;
  height: 1.2rem !important;
  width: 1.2rem !important;
}

.btn.icon-btn.x-small > i {
  font-size: 0.6rem !important;
}

.btn-outlined {
  background-color: transparent;
  border: thin solid !important;

  &.accent {
    background-color: transparent;
    border-color: #2680eb;
    color: #2680eb;

    &:hover {
      background-color: #2680eb;
      border-color: #2680eb;
      color: #fff;
    }
  }
}

.btn.link {
  background-color: transparent;
  border: none;
  color: #2680eb;
  text-decoration: underline;
  padding: 0;

  &:hover {
    background-color: transparent;
    color: #115bb4;
    text-decoration: none;
  }
}

.btn.disabled,
.btn:disabled {
  cursor: default;
  opacity: 0.65;
  pointer-events: none;
}

/* Button group styles */
.btn.btn-group-start {
  border-radius: 4px 0 0 4px;
}

.btn.btn-group-end {
  border-radius: 0 4px 4px 0;
}

.btn.btn-group-vertical-start {
  border-radius: 4px 4px 0 0;
  width: 100%;
}

.btn.btn-group-vertical-end {
  border-radius: 0 0 4px 4px;
  width: 100%;
}

.btn.btn-group.btn-group__btn,
.btn.btn-group-vertical.btn-group__btn {
  border-radius: 0;
  width: 100%;
}
