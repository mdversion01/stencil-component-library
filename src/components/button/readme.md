# pl-button



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type                          | Default  |
| ------------ | ------------- | ----------- | ----------------------------- | -------- |
| `absolute`   | `absolute`    |             | `boolean`                     | `false`  |
| `accordion`  | `accordion`   |             | `boolean`                     | `false`  |
| `active`     | `active`      |             | `boolean`                     | `false`  |
| `ariaLabel`  | `aria-label`  |             | `string`                      | `''`     |
| `block`      | `block`       |             | `boolean`                     | `false`  |
| `bottom`     | `bottom`      |             | `string`                      | `''`     |
| `btnIcon`    | `btn-icon`    |             | `boolean`                     | `false`  |
| `btnText`    | `btn-text`    |             | `string`                      | `''`     |
| `classNames` | `class-names` |             | `string`                      | `''`     |
| `devMode`    | `dev-mode`    |             | `boolean`                     | `false`  |
| `disabled`   | `disabled`    |             | `boolean`                     | `false`  |
| `elevation`  | `elevation`   |             | `string`                      | `''`     |
| `end`        | `end`         |             | `boolean`                     | `false`  |
| `fixed`      | `fixed`       |             | `boolean`                     | `false`  |
| `groupBtn`   | `group-btn`   |             | `boolean`                     | `false`  |
| `iconBtn`    | `icon-btn`    |             | `boolean`                     | `false`  |
| `isOpen`     | `is-open`     |             | `boolean`                     | `false`  |
| `left`       | `left`        |             | `string`                      | `''`     |
| `link`       | `link`        |             | `boolean`                     | `false`  |
| `outlined`   | `outlined`    |             | `boolean`                     | `false`  |
| `pressed`    | `pressed`     |             | `boolean \| string`           | `false`  |
| `right`      | `right`       |             | `string`                      | `''`     |
| `ripple`     | `ripple`      |             | `boolean`                     | `false`  |
| `shape`      | `shape`       |             | `string`                      | `''`     |
| `size`       | `size`        |             | `string`                      | `''`     |
| `slotSide`   | `slot-side`   |             | `"left" \| "none" \| "right"` | `'none'` |
| `start`      | `start`       |             | `boolean`                     | `false`  |
| `stripped`   | `stripped`    |             | `boolean`                     | `false`  |
| `styles`     | `styles`      |             | `string`                      | `''`     |
| `targetId`   | `target-id`   |             | `string`                      | `''`     |
| `text`       | `text`        |             | `boolean`                     | `false`  |
| `textBtn`    | `text-btn`    |             | `boolean`                     | `false`  |
| `titleAttr`  | `title-attr`  |             | `string`                      | `''`     |
| `top`        | `top`         |             | `string`                      | `''`     |
| `url`        | `url`         |             | `string`                      | `''`     |
| `variant`    | `variant`     |             | `string`                      | `''`     |
| `vertical`   | `vertical`    |             | `boolean`                     | `false`  |
| `zIndex`     | `z-index`     |             | `string`                      | `''`     |


## Events

| Event         | Description | Type               |
| ------------- | ----------- | ------------------ |
| `customClick` |             | `CustomEvent<any>` |


## Dependencies

### Used by

 - [accordion-component](../accordion)
 - [accordion-container](../accordion-container)
 - [dropdown-component](../dropdown)

### Graph
```mermaid
graph TD;
  accordion-component --> button-component
  accordion-container --> button-component
  dropdown-component --> button-component
  style button-component fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
