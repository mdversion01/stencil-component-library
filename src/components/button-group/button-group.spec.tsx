// src/components/button-group/button-group.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { ButtonGroup } from './button-group-component';
import { Button } from '../button/button-component';


describe('button-group', () => {
  it('renders horizontal by default with default aria-label', async () => {
    const page = await newSpecPage({
      components: [ButtonGroup],
      html: `<button-group></button-group>`,
    });
    const div = page.root.querySelector('div');
    expect(div.getAttribute('role')).toBe('group');
    expect(div.getAttribute('aria-label')).toBe('Button Group');
    expect(div.classList.contains('btn-group')).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('renders vertical layout when vertical prop is true', async () => {
    const page = await newSpecPage({
      components: [ButtonGroup],
      html: `<button-group vertical></button-group>`,
    });
    const div = page.root.querySelector('div');
    expect(div.classList.contains('btn-group-vertical')).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('applies additional classNames', async () => {
    const page = await newSpecPage({
      components: [ButtonGroup],
      html: `<button-group class-names="custom-class"></button-group>`,
    });
    const div = page.root.querySelector('div');
    expect(div.classList.contains('custom-class')).toBe(true);
    expect(page.root).toMatchSnapshot();
  });

  it('uses custom aria-label if provided', async () => {
    const page = await newSpecPage({
      components: [ButtonGroup],
      html: `<button-group aria-label="Custom Group Label"></button-group>`,
    });
    const div = page.root.querySelector('div');
    expect(div.getAttribute('aria-label')).toBe('Custom Group Label');
    expect(page.root).toMatchSnapshot();
  });

  it('renders slotted buttons', async () => {
    const page = await newSpecPage({
      components: [ButtonGroup],
      html: `
        <button-group>
          <button-component btn-text="One"></button-component>
          <button-component btn-text="Two"></button-component>
        </button-group>
      `,
    });
    expect(page.root.querySelectorAll('button-component').length).toBe(2);
    expect(page.root).toMatchSnapshot();
  });

  it('handles keyboard navigation: Enter and Space keys', async () => {
    const page = await newSpecPage({
      components: [ButtonGroup, Button],
      html: `
        <button-group>
          <button-component btn-text="Click Me"></button-component>
        </button-group>
      `,
    });

    const buttonComp = page.root.querySelector('button-component');

    const clickSpy = jest.fn();
    buttonComp.addEventListener('customClick', clickSpy);

    // Manually emit the customClick for Enter and Space
    buttonComp.dispatchEvent(new CustomEvent('customClick'));
    buttonComp.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();

    expect(clickSpy).toHaveBeenCalledTimes(2);
  });


});
