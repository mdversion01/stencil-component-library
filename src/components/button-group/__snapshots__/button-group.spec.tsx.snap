// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button-group applies additional classNames 1`] = `
<button-group class-names="custom-class">
  <!---->
  <div aria-label="Button Group" class="btn-group custom-class" role="group"></div>
</button-group>
`;

exports[`button-group renders horizontal by default with default aria-label 1`] = `
<button-group>
  <!---->
  <div aria-label="Button Group" class="btn-group" role="group"></div>
</button-group>
`;

exports[`button-group renders slotted buttons 1`] = `
<button-group>
  <!---->
  <div aria-label="Button Group" class="btn-group" role="group">
    <button-component btn-text="One"></button-component>
    <button-component btn-text="Two"></button-component>
  </div>
</button-group>
`;

exports[`button-group renders vertical layout when vertical prop is true 1`] = `
<button-group vertical="">
  <!---->
  <div aria-label="Button Group" class="btn-group-vertical" role="group"></div>
</button-group>
`;

exports[`button-group uses custom aria-label if provided 1`] = `
<button-group aria-label="Custom Group Label">
  <!---->
  <div aria-label="Custom Group Label" class="btn-group" role="group"></div>
</button-group>
`;
